### PR TITLE
AF-2245: Dashbuilder not closing ResultSets and Statements

### PR DESCRIPTION
**Pull request by jesuino**
Bringing back fixes for unclosed statements and result sets

### CODE REVIEW

**Comment by jesuino** [comment]
Please review @dgutierr @domhanak 

Thanks!

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by jesuino** [comment]
jenkins execute full downstream build

**Comment by jesuino** [comment]
jenkins retest this

**Comment by jesuino** [comment]
Jenkins execute full downstream build

File: kie-soup-dataset/kie-soup-dataset-sql/src/main/java/org/dashbuilder/dataprovider/sql/model/Select.java [hunk 1]
```diff
@@ -167,19 +169,35 @@ public String toString() {
...
-        String sql = getSQL();
-        return JDBCUtils.executeQuery(connection, sql);
+    public <R> R fetch(ResultSetConsumer<R> consumer) {
+        try {
+            ResultSetHandler handler = null;
+            try {
+                String sql = getSQL();
+                handler = JDBCUtils.executeQuery(connection, sql);
+                return consumer.consume(handler.getResultSet());
+            }
```
Comment (domhanak) [inline kie-soup-dataset/kie-soup-dataset-sql/src/main/java/org/dashbuilder/dataprovider/sql/model/Select.java]: you are not catching anything here, is that what we want?

File: kie-soup-dataset/kie-soup-dataset-sql/src/main/java/org/dashbuilder/dataprovider/sql/dialect/SQLServerDialect.java [hunk 2]
```diff
@@ -15,6 +15,7 @@
  */
 package org.dashbuilder.dataprovider.sql.dialect;
 
+import java.sql.ResultSet;
 import java.sql.SQLException;
```
Comment (domhanak) [inline kie-soup-dataset/kie-soup-dataset-sql/src/main/java/org/dashbuilder/dataprovider/sql/dialect/SQLServerDialect.java]: Unused now

File: kie-soup-dataset/kie-soup-dataset-sql/src/main/java/org/dashbuilder/dataprovider/sql/dialect/SQLServerDialect.java [hunk 3]
```diff
@@ -122,10 +124,15 @@ public String getSQL(Select select) {
         try {
             // Disable limits & fetch results
             select.limit(0).offset(0);
-            return JDBCUtils.getColumns(select.fetch(), null);
-        }
-        catch (SQLException e) {
-            return Collections.emptyList();
+            return select.fetch(new ResultSetConsumer<List<Column>>() {
```
Comment (domhanak) [inline kie-soup-dataset/kie-soup-dataset-sql/src/main/java/org/dashbuilder/dataprovider/sql/dialect/SQLServerDialect.java]: Can be replaced with lambda

**Comment by domhanak** [comment]
@jesuino Left some comments let me know what you think about them. Also consider adding some tests for touched classes that do not have any yet

File: kie-soup-dataset/kie-soup-dataset-sql/src/main/java/org/dashbuilder/dataprovider/sql/dialect/SQLServerDialect.java (hunk 3 above)
Comment (jesuino) [reply kie-soup-dataset/kie-soup-dataset-sql/src/main/java/org/dashbuilder/dataprovider/sql/dialect/SQLServerDialect.java]: done, thanks @domhanak

**Comment by jesuino** [comment]
Jenkins execute full downstream build

File: kie-soup-dataset/kie-soup-dataset-sql/src/main/java/org/dashbuilder/dataprovider/sql/JDBCUtils.java [hunk 4]
```diff
@@ -114,12 +114,14 @@ public static void execute(Connection connection, String sql) throws SQLExceptio
...
 
-    public static ResultSet executeQuery(Connection connection, String sql) throws SQLException {
+    public static ResultSetHandler executeQuery(Connection connection, String sql) throws SQLException {
         try {
             if (log.isDebugEnabled()) {
                 log.debug(sql);
             }
-            return connection.createStatement().executeQuery(sql);
+            Statement statement = connection.createStatement();
+            ResultSet resultSet = statement.executeQuery(sql);
```
Comment (domhanak) [inline kie-soup-dataset/kie-soup-dataset-sql/src/main/java/org/dashbuilder/dataprovider/sql/JDBCUtils.java]: @jesuino FYI: Sonarcloud yells this is a problem (we are not closing the statement), but imho it is false positive in this particular case so I marked it as won't fix for the future.

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by ederign** [comment]
Jenkins execute full downstream build

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by jesuino** [comment]
The errors does not seem to be related, @ederign  - it is failing some enforcer rule, which is not related to kie-soup project

**Pull request merged**